third_party/move/move-vm/runtime/src/reentrancy_checker.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements a reentrancy checker for dynamic dispatch. Two types of checks
+//! are implemented:
+//!
+//! (1) The resource lock mechanism for closure dispatch, as described in
+//!     [AIP-122](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-112.md).
+//!     In summary, for this mechanism access to any resource is disallowed on reentrancy.
+//! (2) The module lock mechanism for native dispatch as implemented for
+//!     [AIP-73](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-73.md).
+//!     For this mechanism reentrancy via any kind of function call is disallowed.
+//!     This entails (2), so every check failing for (1), also fails in (2). This
+//!     is by the property that resources can only be accessed inside the module
+//!     they are defined in.
+//!
+//! The checker by default operates in mode (1), but allows to enter code
```

Comment:
I think a better semantics would be to keep a feature flag. Such that once the feature flag is activated we will operate in mode (1) for both closure dispatch and native dispatch.

Reply:
I discussed this with OtterSec and they advised against it, because the new mechanism is more liberal, which could lead to issues with existing code.